### PR TITLE
Add pre-decompositions

### PR DESCRIPTION
**Pull request by nopara73**

### CODE REVIEW

File: Sake/Mixer.cs [hunk 1]
```diff
@@ -340,6 +347,70 @@ public IEnumerable<Output> Decompose(IEnumerable<Money> myInputsParam, IEnumerab
...
+            {
+                var remainingVsize = availableVsize;
+                var remaining = myInputSum;
+                List<Output> currentSet = new();
+                while (true)
+                {
+                    var denom = denoms.Where(x => x.EffectiveCost <= remaining && x.EffectiveCost >= (remaining / 3)).RandomElement(Random)
+                        ?? denoms.FirstOrDefault(x => x.EffectiveCost <= remaining);
+
+                    // We can only let this go forward if at least 2 output can be added (denom + potential change)
```
Comment (yahiheb) [inline Sake/Mixer.cs]:
```suggestion
                    // We can only let this go forward if at least 2 outputs can be added (denom + potential change)
```

**Review by yahiheb — CHANGES_REQUESTED** [review]

**Comment by turbolay** [comment]
Unrelated to this PR and maybe I'm wrong but I will take the opportunity to raise the concern, this line seems incomplete:
https://github.com/nopara73/Sake/blob/9beb1bf271478b01aeec6a31a8b7c2ec481c0fa9/Sake/Mixer.cs#L365

Shouldn't this condition be
```diff
    if (denom is null || 
+       (remaining != MinAllowedOutputAmount || remainingVsize > denom.ScriptType.EstimateOutputVsize()) &&
        (remaining < MinAllowedOutputAmount + ChangeFee || remainingVsize < denom.ScriptType.EstimateOutputVsize() + ChangeScriptType.EstimateOutputVsize()))
```
Because if the remaining is exactly the `MinAllowedOutputAmount`, one extra output can be added, without worrying about the `ChangeFee` because there won't be any.

**Comment by nopara73** [comment]
@turbolay I have spent some minutes trying to understand your comment, but I did not fully grasp it just yet. I suspect you're right. Can you make PRs
- to Sake and to WalletWasabi
- to naive and to pre-decompositions (does Lucas decomposition needs something like this?)
- have a decent description of the PRs
- improve the comments above the code modification

**Comment by turbolay** [comment]
Something else here: Maybe I don't exactly understand exactly this line:

https://github.com/nopara73/Sake/blob/9beb1bf271478b01aeec6a31a8b7c2ec481c0fa9/Sake/Mixer.cs#L361-L362

But it seems that in the condition L365 that I quoted in my previous message `MinAllowedOutputAmount` should be replaced by `denom.Amount` otherwise, the remaining can be less than 0 if `denom.Aount > MinAllowedOutputAmount + ChangeFee`

**Comment by nopara73** [comment]
@turbolay if I have issues comprehending your previous comment, then I'll have issues comprehending the comment that builds on top of that 😄 please make the PRs as I suggested so hopefully it'll be easier to understand what you want by looking at the finished work.

**Comment by turbolay** [comment]
(it was not built on top of it, it's a different comment that happens to be on the same line)
I will make the PR today.

**Pull request merged**